File: Flustered Pterodactyl/cocos2dx/MarmeladeGAFPlayer/Examples/demo_features/GafFeatures.cpp
#include "GafFeatures.h"
#include "SimpleAudioEngine.h"

#include "GAFAsset.h"
#include "GAFAnimatedObject.h"

#include "GAFMarmalade.h"

using namespace cocos2d;
using namespace CocosDenshion;
using GAF::GAFAnimatedObject;

GafFeatures::GafFeatures()
:
_asset(NULL),
_objects(NULL),
_externalSpriteContainter(NULL)
{
}
GafFeatures::~GafFeatures()
{
	CC_SAFE_RELEASE(_asset);
	CC_SAFE_RELEASE(_objects);
	if (_externalSpriteContainter)
	{
		GAF::GAFMarmaladeGFX::releaseSpriteContainer(_externalSpriteContainter);
	}
}

CCScene* GafFeatures::scene()
{
    CCScene *scene = CCScene::create();
    GafFeatures *layer = GafFeatures::create();
    scene->addChild(layer);
    return scene;
}

CCMenuItemImage * GafFeatures::addBtn(const char * text, float px, float py, SEL_MenuHandler handler, float k)
{
	CCSize size = CCDirector::sharedDirector()->getWinSize();
	
	CCMenuItemImage *res = CCMenuItemImage::create(
														  "CloseNormal.png",
														  "CloseSelected.png",
														  this,
														  handler );
    res->setPosition( ccp(size.width * px, size.height * py) );
	res->setScale(k);
	
	CCLabelTTF* pLabel = CCLabelTTF::create(text, "Thonburi", 34);
	pLabel->setColor(ccc3(0, 0, 255));
	
	pLabel->setAnchorPoint(CCPoint(1, 0.5));
	pLabel->setScale(k);
    pLabel->setPosition( ccp(res->getPositionX() - res->getContentSize().width * k * 0.5, res->getPositionY()) );
    addChild(pLabel, 100000);
	return res;
}

void GafFeatures::black()
{
	glClearColor(0.0f, 0.0f, 0.0f, 1.0f);
}

void GafFeatures::white()
{
	glClearColor(1.0f, 1.0f, 1.0f, 1.0f);
}

void GafFeatures::gray()
{
	glClearColor(0.5f, 0.5f, 0.5f, 1.0f);
}

void GafFeatures::prevFrame()
{
	int f = frameNumber();
	if (-1 == f)
	{
		return;
	}
	if (f)
	{
		setFrameNumber(f - 1);
	}
}

void GafFeatures::nextFrame()
{
	int f = frameNumber();
	if (-1 == f)
	{
		return;
	}
	if (f != maxFrameNumber())
	{
		setFrameNumber(f + 1);
	}
}

bool GafFeatures::init()
{
	if (_externalSpriteContainter)
	{
		GAF::GAFMarmaladeGFX::releaseSpriteContainer(_externalSpriteContainter);
		_externalSpriteContainter = NULL;
	}
	
	_externalSpriteContainter = GAF::GAFMarmaladeGFX::createSpriteContainer();
	if (!_externalSpriteContainter)
	{
		return false;
	}
	
#if (CC_TARGET_PLATFORM == CC_PLATFORM_IOS)
	GAFAsset::setDesiredCsf(CCDirector::sharedDirector()->getContentScaleFactor());
#endif
	CCSize size = CCDirector::sharedDirector()->getWinSizeInPixels();
	
	float s = size.height / 640;
	if (s > 1) s = 1;
	float dp = 0.13;
	
	CCArray * items = CCArray::create();
	items->addObject(addBtn("Play/Pause", 0.95, 0.95, menu_selector(GafFeatures::playpause), s));
	items->addObject(addBtn("Restart", 0.95, 0.95 -dp, menu_selector(GafFeatures::restart), s));
	items->addObject(addBtn("B", 0.75, 0.95 - dp* 2, menu_selector(GafFeatures::black), s));
	items->addObject(addBtn("W", 0.85, 0.95 - dp* 2, menu_selector(GafFeatures::white), s));
	items->addObject(addBtn("G", 0.95, 0.95 - dp * 2, menu_selector(GafFeatures::gray), s));
	items->addObject(addBtn("-", 0.85, 0.95 - dp * 3, menu_selector(GafFeatures::removeOne), s));
	items->addObject(addBtn("+", 0.95, 0.95 - dp* 3, menu_selector(GafFeatures::addOne), s));
	items->addObject(addBtn("1", 0.75, 0.95 - dp * 4, menu_selector(GafFeatures::set1), s));
	items->addObject(addBtn("5", 0.85, 0.95 - dp* 4, menu_selector(GafFeatures::set5), s));
	items->addObject(addBtn("10", 0.95, 0.95 - dp * 4, menu_selector(GafFeatures::set10), s));
	items->addObject(addBtn("-", 0.85, 0.95 - dp * 5, menu_selector(GafFeatures::prevFrame), s));
	items->addObject(addBtn("+", 0.95, 0.95 - dp* 5, menu_selector(GafFeatures::nextFrame), s));
	items->addObject(addBtn("Full cleanup", 0.95, 0.95 - dp* 6, menu_selector(GafFeatures::cleanup), s));
	items->addObject(addBtn("-", 0.85, 0.95 - dp * 7, menu_selector(GafFeatures::prev_anim), s));
	items->addObject(addBtn("+", 0.95, 0.95 - dp* 7, menu_selector(GafFeatures::next_anim), s));

    CCMenu* pMenu = CCMenu::createWithArray(items);
    pMenu->setPosition(CCPointZero);
    addChild(pMenu, 10000);
	_anim_index = 0;
	
	_jsons.push_back("SampleAnimations/1/1.json");
	_jsons.push_back("SampleAnimations/2/2.json");
	_jsons.push_back("SampleAnimations/3/3.json");
	_jsons.push_back("SampleAnimations/4/4.json");
	
	addObjectsToScene(1);
	black();
	
	setTouchEnabled(true);
	return true;
}

void GafFeatures::ccTouchesBegan(CCSet *pTouches, CCEvent *pEvent)
{
	if (!_objects || !_objects->count())
	{
		return;
	}
	GAFAnimatedObject * node = (GAFAnimatedObject*)_objects->objectAtIndex(0);
	CCTouch * pTouch = (CCTouch*) pTouches->anyObject();
	CCPoint pt = pTouch->getLocation();
	node->setPosition(pt.x, pt.y);
}


void GafFeatures::next_anim()
{
	if (!_jsons.size())
	{
		return;
	}
	cleanup();
	++_anim_index;
	if (_anim_index >= (int)_jsons.size())
	{
		_anim_index = 0;
	}
	addObjectsToScene(1);
		
}

void GafFeatures::prev_anim()
{
	if (!_jsons.size())
	{
		return;
	}
	cleanup();
	--_anim_index;
	if (_anim_index < 0)
	{
		_anim_index = _jsons.size() - 1;
	}
	addObjectsToScene(1);
}

void GafFeatures::restart()
{
	if (!_objects || !_objects->count())
	{
		return;
	}
	GAFAnimatedObject *object = (GAFAnimatedObject *)_objects->objectAtIndex(0);
	object->stop();
	object->start();
}

void GafFeatures::playpause()
{
	if (!_objects || !_objects->count())
	{
		return;
	}
	GAFAnimatedObject *object = (GAFAnimatedObject *)_objects->objectAtIndex(0);
	

	if (object->isAnimationRunning())
	{
		object->pause();
	}
	else
	{
		object->resume();
	}

}

int GafFeatures::maxFrameNumber()
{
	if (!_objects || !_objects->count())
	{
		return -1;
	}
	GAFAnimatedObject *object = (GAFAnimatedObject *)_objects->objectAtIndex(0);
	return object->totalFrameCount();	
}

void GafFeatures::setFrameNumber(int aFrameNumber)
{
	if (!_objects || !_objects->count())
	{
		return;
	}
	GAFAnimatedObject *object = (GAFAnimatedObject *)_objects->objectAtIndex(0);
	object->setFrame(aFrameNumber);	
}

int GafFeatures::frameNumber()
{
	if (!_objects || !_objects->count())
	{
		return -1;
	}
	GAFAnimatedObject *object = (GAFAnimatedObject *)_objects->objectAtIndex(0);
	return object->currentFrameIndex();
}


void GafFeatures::set1()
{
	set(1);
}

void GafFeatures::set5()
{
	set(5);
}

void GafFeatures::set10()
{
	set(10);
}

void GafFeatures::set(int n)
{
	if (!_objects)
	{
		_objects = GAF::CCArray::create();
		_objects->retain();
	}
	int c = _objects->count();
	if (c == n)
	{
		return;
	}
	if (n > c)
	{
		addObjectsToScene(n - c);
	}
	else
	{
		removeFromScene(c - n);
	}
}

void GafFeatures::addOne()
{
	addObjectsToScene(1);
}

void GafFeatures::removeOne()
{
	removeFromScene(1);
}

void GafFeatures::cleanup()
{
	CC_SAFE_RELEASE_NULL(_asset);
	if (!_objects)
	{
		return;
	}
	removeFromScene(_objects->count());
	CC_SAFE_RELEASE_NULL(_objects);
}

void GafFeatures::removeFromScene(int aCount)
{
	if (!_objects || !aCount)
	{
		return;
	}
	if (aCount > int(_objects->count()))
	{
		aCount = _objects->count();
	}
	for (int i = 0; i < aCount; ++i)
	{
		GAFAnimatedObject *obj = (GAFAnimatedObject *) _objects->lastObject();
		GAF::GAFMarmaladeGFX::sceneRemoveSpriteContainer(this, obj->getExternalPointer());
		_objects->removeLastObject();
	}
}

void GafFeatures::addObjectsToScene(int aCount)
{
	if (!aCount)
	{
		return;
	}
	if (!_asset)
	{		
		_asset =  GAFAsset::create(_jsons[_anim_index]);
		CC_SAFE_RETAIN(_asset);
	}
	if (!_objects)
	{
		_objects = GAF::CCArray::create();
		_objects->retain();
	}
	if (_asset)
	{
		int initialCount = _objects->count();
		for (int i = initialCount; i < initialCount + aCount; ++i)
		{
			GAFAnimatedObject *object = _asset->createObjectAndRun(true);
			GAF::GAFMarmaladeGFX::setSpriteContainerZorder(object->getExternalPointer(), 100 * i);
			GAF::GAFMarmaladeGFX::sceneAddSpriteContainer(this, object->getExternalPointer());


			float s = 768 / 640;
			if (s > 1) s = 1;
			object->setScale(s);			

			object->setPosition(1024 * (0.4 + 0.05 * i), 768 * 0.6);
			_objects->addObject(object);
			
			// will work only if animation has sequence
			object->playSequence("walk", true);

		}
	}
}

File: BIB/cocos2dx/gaf_runtimes/cocos2d-iphone/demo_features/GafFeatures.m
#import "GafFeatures.h"

#import "AppDelegate.h"
#import "GAFAsset.h"
#import "GAFAnimatedObject.h"

@implementation GafFeatures

+(CCScene *) scene
{
	CCScene *scene = [CCScene node];
	GafFeatures *layer = [GafFeatures node];
	[scene addChild: layer];
	return scene;
}

-(id) init
{
	if( (self=[super init]) )
	{
		CGSize size = [[CCDirector sharedDirector] winSizeInPixels];
		float s = size.height / 640;
		if (s > 1) s = 1;
		float dp = 0.13;
		
		NSMutableArray * items = [[[NSMutableArray alloc] init] autorelease];

		[items addObject: [self addBtn:@"Play/Pause" px:0.95 py:0.95 handler:@selector(playpause) k:s]];
		[items addObject: [self addBtn:@"Restart" px:0.95 py:0.95-dp handler:@selector(restart) k:s]];
		[items addObject: [self addBtn:@"B" px:0.75 py:0.95- dp* 2 handler:@selector(black) k:s]];
		[items addObject: [self addBtn:@"W" px:0.85 py:0.95- dp* 2 handler:@selector(white) k:s]];
		[items addObject: [self addBtn:@"G" px:0.95 py:0.95- dp* 2 handler:@selector(gray) k:s]];
		[items addObject: [self addBtn:@"-" px:0.85 py:0.95- dp* 3 handler:@selector(removeOne) k:s]];
		[items addObject: [self addBtn:@"+" px:0.95 py:0.95- dp* 3 handler:@selector(addOne) k:s]];
		[items addObject: [self addBtn:@"1" px:0.75 py:0.95- dp* 4 handler:@selector(set1) k:s]];
		[items addObject: [self addBtn:@"5" px:0.85 py:0.95- dp* 4 handler:@selector(set5) k:s]];
		[items addObject: [self addBtn:@"10" px:0.95 py:0.95- dp* 4 handler:@selector(set10) k:s]];
		[items addObject: [self addBtn:@"-" px:0.85 py:0.95- dp* 5 handler:@selector(prevFrame) k:s]];
		[items addObject: [self addBtn:@"+" px:0.95 py:0.95- dp* 5 handler:@selector(nextFrame) k:s]];
		[items addObject: [self addBtn:@"Full cleanup" px:0.95 py:0.95- dp* 6 handler:@selector(cleanup) k:s]];
		[items addObject: [self addBtn:@"-" px:0.85 py:0.95- dp* 7 handler:@selector(prev_anim) k:s]];
		[items addObject: [self addBtn:@"+" px:0.95 py:0.95- dp* 7 handler:@selector(next_anim) k:s]];
		
		CCMenu * pMenu = [CCMenu menuWithArray:items];
		pMenu.position = CGPointZero;
		[self addChild:pMenu z:10000];		
		_anim_index = 0;
		_jsons = [[NSMutableArray alloc] init];
		
		[_jsons addObject:@"SampleAnimations/enemy_slime/1HNYAK0T5VWH.json"];
		[_jsons addObject:@"SampleAnimations/fireman/0619WC0I55O0.json"];		
		[_jsons addObject:@"SampleAnimations/simpleMovement/simpleMovement.json"];
		[_jsons addObject:@"SampleAnimations/blurTest/blurTest.json"];
		[_jsons addObject:@"SampleAnimations/moveScaleRotate/moveScaleRotate.json"];
		[_jsons addObject:@"SampleAnimations/simpleMovementWithColor/simpleMovementWithColor.json"];
		[_jsons addObject:@"SampleAnimations/newMaskTest/newMaskTest.json"];
		[_jsons addObject:@"SampleAnimations/masksTest/masksTest.json"];
#if 0	
		[_jsons addObject:@"SampleAnimations/fireman/15RA0B0KSJ4N.json"];
		[_jsons addObject:@"SampleAnimations/multiAtlasesTest/1BPM2302S80NAsset.json"];
		[_jsons addObject:@"SampleAnimations/named/1BPM211BI6SRAsset.json"];

#endif
		[self addObjectsToScene:1];
		[self white];
		[self setTouchEnabled:YES];

	}
	return self;
}

- (CCMenuItemImage*) addBtn:(NSString*) text px:(float)px py:(float)py  handler:(SEL)handler k:(float)k
{
	CGSize size = [[CCDirector sharedDirector] winSize];
	
	CCMenuItemImage *res = [CCMenuItemImage itemWithNormalImage:@"CloseNormal.png" selectedImage:@"CloseSelected.png" target:self selector:handler];

    res.position = ccp(size.width * px, size.height * py);
	res.scale = k;
	
	CCLabelTTF* pLabel = [CCLabelTTF labelWithString:text fontName:@"Thonburi" fontSize:34];
	pLabel.color = ccc3(0, 0, 255);
	
	pLabel.anchorPoint = ccp(1, 0.5);
	pLabel.scale = k;
    pLabel.position = ccp(res.position.x - [res contentSize].width * k * 0.5, res.position.y) ;
	[self addChild:pLabel z:100000];
	
	return res;
}

-(void) black
{
	glClearColor(0.0f, 0.0f, 0.0f, 1.0f);
}

-(void) white
{
	glClearColor(1.0f, 1.0f, 1.0f, 1.0f);
}

-(void) gray
{
	glClearColor(0.5f, 0.5f, 0.5f, 1.0f);
}


-(void) prevFrame
{
	int f = [self frameNumber];
	if (-1 == f)
	{
		return;
	}
	if (f != [self maxFrameNumber])
	{
		[self setFrameNumber:f - 1];
	}
}

-(void) nextFrame
{
	int f = [self frameNumber];
	if (-1 == f)
	{
		return;
	}
	if (f != [self maxFrameNumber])
	{
		[self setFrameNumber:f + 1];
	}
}

-(void) addOne
{
	[self addObjectsToScene:1];
}

-(void) removeOne
{
	[self removeFromScene:1];
}

-(void) set:(int)n
{
	if (_objects == nil)
	{
		_objects = [[NSMutableArray alloc] init];
	}
	int c = [_objects count];
	if (c == n)
	{
		return;
	}
	if (n > c)
	{
		[self addObjectsToScene:n - c];
	}
	else
	{
		[self removeFromScene:c - n];
	}
}

- (void)ccTouchesBegan:(NSSet *)touches withEvent:(UIEvent *)event
{
	if (_objects == nil || 0 == [_objects count])
	{
		return;
	}
	
	CCNode * node = (CCNode*) [_objects objectAtIndex:0];
	UITouch * pTouch = (UITouch *)[touches anyObject];
	CGPoint pt = [pTouch locationInView: [CCDirector sharedDirector].view ];
	pt.y = [CCDirector sharedDirector].winSize.height - pt.y;
	node.position = pt;
}

-(void) addObjectsToScene:(int) aCount
{
	if (_asset == nil)
	{
		_asset = [[GAFAsset create:[_jsons objectAtIndex:_anim_index]] retain];
	}
	if (_objects == nil)
	{
		_objects = [[NSMutableArray alloc] init];
	}
	if (_asset != nil)
	{
		int initialCount = [_objects count];
		for (int i = initialCount; i < initialCount + aCount; ++i)
		{
			GAFAnimatedObject *object = [_asset createObjectAndRun:YES];
			object.zOrder = 100 * i;
			[self addChild:object];
			CGSize winSize = [[CCDirector sharedDirector] winSize];
			object.position = ccp(winSize.width * (0.5 + 0.05 * i), winSize.height * 0.5);
			[_objects addObject:object];
			// will work only if animation has sequence
			[object setSequenceWithName:@"walk" looped:YES];
		}
	}
}

-(void) removeFromScene:(int) aCount
{
	if (_objects == nil)
	{
		return;
	}
	if (aCount > [_objects count])
	{
		aCount = [_objects count];
	}
	for (int i = 0; i < aCount; ++i)
	{
		GAFAnimatedObject *obj = (GAFAnimatedObject *)[_objects lastObject];
		[obj removeFromParentAndCleanup:YES];
		[_objects removeLastObject];
	}
}


-(void) set1
{
	[self set:1];
}

-(void) set5
{
	[self set:5];
}

-(void) set10
{
	[self set:10];
}
//-(void) set(int n);
-(void) restart
{
	if (_objects == nil || 0 == [_objects count])
	{
		return;
	}
	GAFAnimatedObject *object = (GAFAnimatedObject *)[_objects objectAtIndex:0];
	[object stop];
	[object start];
}

-(void) playpause
{
	if (_objects == nil || 0 == [_objects count])
	{
		return;
	}
	GAFAnimatedObject *object = (GAFAnimatedObject *)[_objects objectAtIndex:0];
	
	if (object.isRunning)
	{
		[object pause];
	}
	else
	{
		[object resume];
	}
}

-(void) cleanup
{
	[_asset release];
	_asset = nil;
	if (_objects == nil || 0 == [_objects count])
	{
		return;
	}
	[self removeFromScene:[_objects count]];
	
	[_objects release];
	_objects = nil;
}

-(int)  maxFrameNumber
{
	if (_objects == nil || 0 == [_objects count])
	{
		return -1;
	}
	GAFAnimatedObject *object = (GAFAnimatedObject *)[_objects objectAtIndex:0];
	return object.totalFrameCount;
}

-(void) setFrameNumber:(int) aFrameNumber
{
	if (_objects == nil || 0 == [_objects count])
	{
		return;
	}
	GAFAnimatedObject *object = (GAFAnimatedObject *)[_objects objectAtIndex:0];
	[object setFrame:aFrameNumber];
}

-(int) frameNumber
{
	if (_objects == nil || 0 == [_objects count])
	{
		return -1;
	}
	GAFAnimatedObject *object = (GAFAnimatedObject *)[_objects objectAtIndex:0];
	return object.currentFrameIndex;
}

-(void) next_anim
{
	if (_jsons == nil || 0 == [_jsons count])
	{
		return;
	}
	[self cleanup];
	++_anim_index;
	if (_anim_index >= [_jsons count])
	{
		_anim_index = 0;
	}
	[self addObjectsToScene:1];
}

-(void) prev_anim
{
	if (_jsons == nil || 0 == [_jsons count])
	{
		return;
	}
	[self cleanup];
	--_anim_index;
	if (_anim_index < 0)
	{
		_anim_index = [_jsons count] - 1;
	}
	[self addObjectsToScene:1];
}

// on "dealloc" you need to release all your retained objects
- (void) dealloc
{
	[_asset release];
	[_objects release];
	[_jsons release];
	[super dealloc];
}

@end
